File: js/src/a2a.ts
/**
 * A2A (agent-to-agent) — OUTBOUND: call remote A2A agents.
 *
 * A remote agent publishes an Agent Card at `/.well-known/agent-card.json`
 * describing its skills. Each advertised skill becomes a uniform `Tool`
 * (source:"a2a") whose `execute` performs one JSON-RPC `SendMessage`
 * (non-blocking) to get a Task id, then polls `GetTask` until the Task reaches
 * a terminal state — mapping the Task's artifact/message text to a ToolResult.
 *
 * This is a genuine subset of real A2A (JSON-RPC 2.0 over one HTTP POST
 * endpoint), so a toolnexus agent interoperates with real A2A peers. See
 * ../../SPEC.md and openspec/changes/add-a2a-agents. Reuses the `${ENV}` header
 * expansion from mcp.ts; secrets live in the environment and are never logged.
 */
import { randomUUID } from "node:crypto"
import { type JSONSchema, type Tool, type ToolContext, type ToolResult, sanitize } from "./types.js"
import { expandEnvHeaders } from "./mcp.js"

const DEFAULT_TIMEOUT = 300_000
const DEFAULT_POLL_EVERY = 1_000

/** Terminal A2A task states — polling stops once one of these is reached. */
const TERMINAL = new Set(["completed", "failed", "canceled"])

// ---------------------------------------------------------------------------
// Wire types (the A2A subset we speak — verified against a2a-python).
// ---------------------------------------------------------------------------

export type A2ATaskState = "submitted" | "working" | "completed" | "failed" | "canceled" | string

export interface A2APart {
  kind: string
  text?: string
  [k: string]: unknown
}

export interface A2AMessage {
  role: string
  messageId?: string
  parts?: A2APart[]
  [k: string]: unknown
}

export interface A2AArtifact {
  parts?: A2APart[]
  [k: string]: unknown
}

export interface A2ATaskStatus {
  state: A2ATaskState
  message?: A2AMessage
  [k: string]: unknown
}

export interface A2ATask {
  id: string
  /** Groups a peer's turns into one conversation (A2A contextId). */
  contextId?: string
  status: A2ATaskStatus
  artifacts?: A2AArtifact[]
  history?: A2AMessage[]
  [k: string]: unknown
}

export interface A2ASkill {
  id?: string
  name?: string
  description?: string
  tags?: string[]
  examples?: string[]
  [k: string]: unknown
}

export interface AgentCard {
  name: string
  description?: string
  version?: string
  protocolVersion?: string
  capabilities?: { streaming?: boolean; pushNotifications?: boolean; [k: string]: unknown }
  defaultInputModes?: string[]
  defaultOutputModes?: string[]
  skills?: A2ASkill[]
  /** The JSON-RPC endpoint the agent listens on. */
  url?: string
  [k: string]: unknown
}

// ---------------------------------------------------------------------------
// Agent descriptor + factory.
// ---------------------------------------------------------------------------

export interface Agent {
  /** URL of the Agent Card (`/.well-known/agent-card.json`). */
  card: string
  /** `${ENV_VAR}` header values are expanded at call time and never logged. */
  headers?: Record<string, string>
  /** Overall poll budget in ms (default 300000). */
  timeout?: number
  /** Interval between `GetTask` polls in ms (default 1000). */
  pollEvery?: number
}

/** Build an Agent descriptor pointing at a remote Agent Card URL. */
export function agent(opts: Agent): Agent {
  return { card: opts.card, headers: opts.headers, timeout: opts.timeout, pollEvery: opts.pollEvery }
}

/** Normalize either an Agent descriptor or a bare card URL into an Agent. */
function toAgent(agentOrCardUrl: Agent | string, opts?: Omit<Agent, "card">): Agent {
  if (typeof agentOrCardUrl === "string") return agent({ card: agentOrCardUrl, ...opts })
  return agentOrCardUrl
}

// ---------------------------------------------------------------------------
// Config parsing — an `agents` block mirroring `mcpServers` (§2).
// ---------------------------------------------------------------------------

export interface AgentConfig {
  card: string
  headers?: Record<string, string>
  timeout?: number
  pollEvery?: number
  enabled?: boolean
  disabled?: boolean
}

export type AgentsConfig = Record<string, AgentConfig>

/** MCP `isEnabled` precedence: `disabled:true` wins, then `enabled:false`. */
function isEnabled(cfg: AgentConfig): boolean {
  if (cfg.disabled === true) return false
  if (cfg.enabled === false) return false
  return true
}

/**
 * Parse an `agents` block (Record<name, AgentConfig>) into Agent descriptors,
 * skipping disabled entries. The config key is just an identifier — a tool's
 * name prefix comes from the fetched card's `name`, not the key.
 */
export function parseAgentsConfig(block: AgentsConfig | undefined): Agent[] {
  if (!block) return []
  const out: Agent[] = []
  for (const cfg of Object.values(block)) {
    if (!cfg || typeof cfg.card !== "string") continue
    if (!isEnabled(cfg)) continue
    out.push(agent({ card: cfg.card, headers: cfg.headers, timeout: cfg.timeout, pollEvery: cfg.pollEvery }))
  }
  return out
}

// ---------------------------------------------------------------------------
// JSON-RPC transport (single POST) + card fetch. Reuses httpTool's ${ENV}
// header expansion, timeout handling, and non-2xx → error mapping.
// ---------------------------------------------------------------------------

/** POST one JSON-RPC 2.0 request and return `result` (throws on error/non-2xx). */
async function jsonRpc(
  endpoint: string,
  method: string,
  params: unknown,
  headers: Record<string, string>,
  timeout: number,
  signal?: AbortSignal,
): Promise<unknown> {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), timeout)
  if (signal) signal.addEventListener("abort", () => controller.abort(), { once: true })
  try {
    const res = await fetch(endpoint, {
      method: "POST",
      headers: { "Content-Type": "application/json", ...headers },
      body: JSON.stringify({ jsonrpc: "2.0", id: randomUUID(), method, params }),
      signal: controller.signal,
    })
    const text = await res.text()
    if (!res.ok) throw new Error(`HTTP ${res.status}: ${text}`)
    const payload = safeParse(text) as { result?: unknown; error?: { message?: string; code?: number } }
    if (payload && payload.error) {
      throw new Error(payload.error.message ?? `JSON-RPC error ${payload.error.code ?? ""}`.trim())
    }
    return payload?.result
  } finally {
    clearTimeout(timer)
  }
}

/** Fetch and parse the Agent Card at its URL (GET). */
async function fetchCard(
  cardUrl: string,
  headers: Record<string, string>,
  timeout: number,
  signal?: AbortSignal,
): Promise<AgentCard> {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), timeout)
  if (signal) signal.addEventListener("abort", () => controller.abort(), { once: true })
  try {
    const res = await fetch(cardUrl, { method: "GET", headers, signal: controller.signal })
    const text = await res.text()
    if (!res.ok) throw new Error(`HTTP ${res.status}: ${text}`)
    return JSON.parse(text) as AgentCard
  } finally {
    clearTimeout(timer)
  }
}

function safeParse(text: string): unknown {
  try {
    return JSON.parse(text)
  } catch {
    return undefined
  }
}

/** Abortable sleep — resolves early (without error) if `signal` fires. */
function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve) => {
    if (signal?.aborted) return resolve()
    const onAbort = () => {
      clearTimeout(timer)
      resolve()
    }
    const timer = setTimeout(() => {
      signal?.removeEventListener("abort", onAbort)
      resolve()
    }, ms)
    signal?.addEventListener("abort", onAbort, { once: true })
  })
}

/** Concatenate a Task's text output: artifact text parts, else last agent message. */
function extractOutput(task: A2ATask): string {
  const parts: string[] = []
  for (const artifact of task.artifacts ?? []) {
    for (const part of artifact.parts ?? []) {
      if (part.kind === "text" && typeof part.text === "string") parts.push(part.text)
    }
  }
  if (parts.length > 0) return parts.join("\n")
  // Fallback: the last agent message in history.
  const history = task.history ?? []
  for (let i = history.length - 1; i >= 0; i--) {
    const msg = history[i]
    if (msg.role !== "agent") continue
    const text = (msg.parts ?? [])
      .filter((p) => p.kind === "text" && typeof p.text === "string")
      .map((p) => p.text)
      .join("\n")
    if (text) return text
  }
  return ""
}

/** Text of a Task's status.message (used for failed/canceled error output). */
function statusMessageText(task: A2ATask): string {
  const msg = task.status?.message
  if (!msg) return ""
  return (msg.parts ?? [])
    .filter((p) => p.kind === "text" && typeof p.text === "string")
    .map((p) => p.text)
    .join("\n")
}

// ---------------------------------------------------------------------------
// Skill → Tool.
// ---------------------------------------------------------------------------

const TASK_SCHEMA: JSONSchema = {
  type: "object",
  properties: { task: { type: "string", description: "The task to send to the agent, in natural language." } },
  required: ["task"],
  additionalProperties: false,
}

/** Build one `source:"a2a"` Tool for a single advertised skill of an agent. */
function skillTool(agentName: string, endpoint: string, skill: A2ASkill, ag: Agent): Tool {
  const skillId = skill.id ?? skill.name ?? ""
  const name = sanitize(agentName) + "_" + sanitize(skillId)
  const timeout = ag.timeout ?? DEFAULT_TIMEOUT
  const pollEvery = ag.pollEvery ?? DEFAULT_POLL_EVERY
  return {
    name,
    description: skill.description ?? skill.name ?? skillId,
    inputSchema: TASK_SCHEMA,
    source: "a2a",
    async execute(args: Record<string, unknown>, ctx?: ToolContext): Promise<ToolResult> {
      const start = Date.now()
      const budget = ctx?.timeout ?? timeout
      const headers = expandEnvHeaders(ag.headers) ?? {}
      const taskText = String(args?.task ?? "")
      let polls = 0
      let taskId = ""
      let state: A2ATaskState = "submitted"
      const meta = () => ({ agent: agentName, taskId, state, polls, ms: Date.now() - start })
      try {
        // 1. SendMessage (non-blocking) → a submitted Task.
        let task = (await jsonRpc(
          endpoint,
          "SendMessage",
          {
            message: { role: "user", messageId: randomUUID(), parts: [{ kind: "text", text: taskText }] },
            configuration: { blocking: false },
          },
          headers,
          budget,
          ctx?.signal,
        )) as A2ATask
        taskId = task?.id ?? ""
        state = task?.status?.state ?? "submitted"

        // 2. Poll GetTask until terminal / timeout / cancel.
        while (!TERMINAL.has(state)) {
          if (ctx?.signal?.aborted) {
            state = "canceled"
            return { output: `A2A task ${taskId} canceled`, isError: true, metadata: meta() }
          }
          if (Date.now() - start >= budget) {
            return {
              output: `A2A task ${taskId} timed out after ${budget}ms (state=${state})`,
              isError: true,
              metadata: meta(),
            }
          }
          await sleep(pollEvery, ctx?.signal)
          // Cancelled during the wait → stop before another GetTask.
          if (ctx?.signal?.aborted) {
            state = "canceled"
            return { output: `A2A task ${taskId} canceled`, isError: true, metadata: meta() }
          }
          task = (await jsonRpc(endpoint, "GetTask", { id: taskId }, headers, budget, ctx?.signal)) as A2ATask
          polls++
          state = task?.status?.state ?? state
        }

        // 3. Map the terminal Task → ToolResult.
        if (state === "completed") {
          return { output: extractOutput(task), isError: false, metadata: meta() }
        }
        const detail = statusMessageText(task)
        return {
          output: `A2A task ${taskId} ${state}${detail ? `: ${detail}` : ""}`,
          isError: true,
          metadata: meta(),
        }
      } catch (e) {
        return { output: e instanceof Error ? e.message : String(e), isError: true, metadata: meta() }
      }
    },
  }
}

/**
 * Resolve an Agent to its tools: fetch the card, read `skills[]`, and produce
 * one Tool per skill. The agent name prefix comes from the card's `name`.
 */
export async function agentTools(ag: Agent): Promise<Tool[]> {
  const timeout = ag.timeout ?? DEFAULT_TIMEOUT
  const headers = expandEnvHeaders(ag.headers) ?? {}
  const card = await fetchCard(ag.card, headers, timeout)
  const agentName = card.name ?? "agent"
  // The card's `url` is the JSON-RPC endpoint; fall back to the card origin.
  const endpoint = card.url ?? new URL(ag.card).origin
  return (card.skills ?? []).map((skill) => skillTool(agentName, endpoint, skill, ag))
}

File: golang/a2a.go
// A2A (agent-to-agent) — OUTBOUND: call remote A2A agents.
//
// A remote agent publishes an Agent Card at `/.well-known/agent-card.json`
// describing its skills. Each advertised skill becomes a uniform Tool
// (source:"a2a") whose Execute performs one JSON-RPC SendMessage (non-blocking)
// to get a Task id, then polls GetTask until the Task reaches a terminal state —
// mapping the Task's artifact/message text to a ToolResult.
//
// This is a genuine subset of real A2A (JSON-RPC 2.0 over one HTTP POST
// endpoint), so a toolnexus agent interoperates with real A2A peers. See
// ../SPEC.md §7A and openspec/changes/add-a2a-agents. Reuses the ${ENV} header
// expansion from http.go; secrets live in the environment and are never logged.

package toolnexus

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
)

const (
	a2aDefaultTimeout   = 300_000 * time.Millisecond
	a2aDefaultPollEvery = 1_000 * time.Millisecond
)

// terminal reports whether a task state stops the poll loop.
func terminal(state string) bool {
	return state == "completed" || state == "failed" || state == "canceled"
}

// ---------------------------------------------------------------------------
// Wire types (the A2A subset we speak — verified against a2a-python).
// ---------------------------------------------------------------------------

// A2APart is one content part of a message or artifact. Text is a pointer so a
// missing `text` (nil) is distinguished from an explicit empty string.
type A2APart struct {
	Kind string  `json:"kind"`
	Text *string `json:"text"`
}

// A2AMessage is a role-tagged list of parts.
type A2AMessage struct {
	Role      string    `json:"role"`
	MessageID string    `json:"messageId,omitempty"`
	Parts     []A2APart `json:"parts,omitempty"`
}

// A2AArtifact is a task output artifact.
type A2AArtifact struct {
	// ArtifactID identifies the artifact (set on served completed tasks).
	ArtifactID string    `json:"artifactId,omitempty"`
	Parts      []A2APart `json:"parts,omitempty"`
}

// A2ATaskStatus carries the task state plus an optional status message.
type A2ATaskStatus struct {
	State   string      `json:"state"`
	Message *A2AMessage `json:"message,omitempty"`
}

// A2ATask is the unit of work returned by SendMessage / GetTask.
type A2ATask struct {
	ID string `json:"id"`
	// ContextID groups a peer's turns into one conversation (A2A contextId).
	ContextID string        `json:"contextId,omitempty"`
	Status    A2ATaskStatus `json:"status"`
	Artifacts []A2AArtifact `json:"artifacts,omitempty"`
	History   []A2AMessage  `json:"history,omitempty"`
}

// A2ASkill is one advertised skill from an Agent Card.
type A2ASkill struct {
	ID          string   `json:"id,omitempty"`
	Name        string   `json:"name,omitempty"`
	Description string   `json:"description,omitempty"`
	Tags        []string `json:"tags,omitempty"`
	Examples    []string `json:"examples,omitempty"`
}

// AgentCard is the descriptor published at /.well-known/agent-card.json.
type AgentCard struct {
	Name               string         `json:"name"`
	Description        string         `json:"description,omitempty"`
	Version            string         `json:"version,omitempty"`
	ProtocolVersion    string         `json:"protocolVersion,omitempty"`
	Capabilities       map[string]any `json:"capabilities,omitempty"`
	DefaultInputModes  []string       `json:"defaultInputModes,omitempty"`
	DefaultOutputModes []string       `json:"defaultOutputModes,omitempty"`
	Skills             []A2ASkill     `json:"skills,omitempty"`
	// URL is the JSON-RPC endpoint the agent listens on.
	URL string `json:"url,omitempty"`
}

// ---------------------------------------------------------------------------
// Agent descriptor.
// ---------------------------------------------------------------------------

// Agent points at a remote Agent Card. Headers support ${ENV_VAR} expansion at
// call time and are never logged.
type Agent struct {
	// Card is the URL of the Agent Card (/.well-known/agent-card.json).
	Card string `json:"card"`
	// Headers are static request headers; ${ENV_VAR} values expand at call time.
	Headers map[string]string `json:"headers,omitempty"`
	// Timeout is the overall poll budget in ms (default 300000).
	Timeout int `json:"timeout,omitempty"`
	// PollEvery is the interval between GetTask polls in ms (default 1000).
	PollEvery int `json:"pollEvery,omitempty"`
}

func (ag Agent) timeout() time.Duration {
	if ag.Timeout > 0 {
		return time.Duration(ag.Timeout) * time.Millisecond
	}
	return a2aDefaultTimeout
}

func (ag Agent) pollEvery() time.Duration {
	if ag.PollEvery > 0 {
		return time.Duration(ag.PollEvery) * time.Millisecond
	}
	return a2aDefaultPollEvery
}

// ---------------------------------------------------------------------------
// Config parsing — an `agents` block mirroring `mcpServers` (§2).
// ---------------------------------------------------------------------------

// AgentConfig is a single entry in a config-file `agents` block.
type AgentConfig struct {
	Card      string            `json:"card"`
	Headers   map[string]string `json:"headers"`
	Timeout   int               `json:"timeout"`
	PollEvery int               `json:"pollEvery"`
	Enabled   *bool             `json:"enabled"`
	Disabled  *bool             `json:"disabled"`
}

// AgentsConfig maps an identifier -> AgentConfig.
type AgentsConfig map[string]AgentConfig

// agentConfigEnabled applies MCP isEnabled precedence: disabled:true wins, then
// enabled:false.
func agentConfigEnabled(cfg AgentConfig) bool {
	if cfg.Disabled != nil && *cfg.Disabled {
		return false
	}
	if cfg.Enabled != nil && !*cfg.Enabled {
		return false
	}
	return true
}

// ParseAgentsConfig turns an `agents` block into Agent descriptors, skipping
// disabled entries. The config key is only an identifier — a tool's name prefix
// comes from the fetched card's name, not the key.
func ParseAgentsConfig(block AgentsConfig) []Agent {
	if block == nil {
		return nil
	}
	var out []Agent
	for _, cfg := range block {
		if cfg.Card == "" {
			continue
		}
		if !agentConfigEnabled(cfg) {
			continue
		}
		out = append(out, Agent{
			Card:      cfg.Card,
			Headers:   cfg.Headers,
			Timeout:   cfg.Timeout,
			PollEvery: cfg.PollEvery,
		})
	}
	return out
}

// ---------------------------------------------------------------------------
// JSON-RPC transport (single POST) + card fetch. Reuses http.go's ${ENV} header
// expansion, context timeout, and non-2xx → error mapping.
// ---------------------------------------------------------------------------

// jsonRPC POSTs one JSON-RPC 2.0 request and returns the raw `result` (error on
// error/non-2xx). budget bounds this single request.
func jsonRPC(parent context.Context, endpoint, method string, params any, headers map[string]string, budget time.Duration) (json.RawMessage, error) {
	ctx, cancel := context.WithTimeout(parent, budget)
	defer cancel()

	envelope := map[string]any{
		"jsonrpc": "2.0",
		"id":      uuid.NewString(),
		"method":  method,
		"params":  params,
	}
	body, err := json.Marshal(envelope)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(raw))
	}
	// safeParse: a body that is not valid JSON yields a nil result (no throw).
	var payload struct {
		Result json.RawMessage `json:"result"`
		Error  *struct {
			Message string `json:"message"`
			Code    *int   `json:"code"`
		} `json:"error"`
	}
	_ = json.Unmarshal(raw, &payload)
	if payload.Error != nil {
		msg := payload.Error.Message
		if msg == "" {
			code := ""
			if payload.Error.Code != nil {
				code = strconv.Itoa(*payload.Error.Code)
			}
			msg = strings.TrimSpace("JSON-RPC error " + code)
		}
		return nil, errors.New(msg)
	}
	return payload.Result, nil
}

// fetchCard GETs and parses the Agent Card at its URL.
func fetchCard(parent context.Context, cardURL string, headers map[string]string, budget time.Duration) (*AgentCard, error) {
	ctx, cancel := context.WithTimeout(parent, budget)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, cardURL, nil)
	if err != nil {
		return nil, err
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(raw))
	}
	var card AgentCard
	if err := json.Unmarshal(raw, &card); err != nil {
		return nil, err
	}
	return &card, nil
}

// aborted reports whether the context is already done (cancel/timeout).
func aborted(ctx context.Context) bool {
	if ctx == nil {
		return false
	}
	select {
	case <-ctx.Done():
		return true
	default:
		return false
	}
}

// sleepCtx sleeps for d, returning early (without error) if ctx is done.
func sleepCtx(ctx context.Context, d time.Duration) {
	if ctx == nil {
		time.Sleep(d)
		return
	}
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-ctx.Done():
	case <-t.C:
	}
}

// extractOutput concatenates a task's text output: artifact text parts, else the
// last agent message.
func extractOutput(task A2ATask) string {
	var parts []string
	for _, artifact := range task.Artifacts {
		for _, p := range artifact.Parts {
			if p.Kind == "text" && p.Text != nil {
				parts = append(parts, *p.Text)
			}
		}
	}
	if len(parts) > 0 {
		return strings.Join(parts, "\n")
	}
	// Fallback: the last agent message in history.
	for i := len(task.History) - 1; i >= 0; i-- {
		msg := task.History[i]
		if msg.Role != "agent" {
			continue
		}
		var texts []string
		for _, p := range msg.Parts {
			if p.Kind == "text" && p.Text != nil {
				texts = append(texts, *p.Text)
			}
		}
		if text := strings.Join(texts, "\n"); text != "" {
			return text
		}
	}
	return ""
}

// statusMessageText returns the text of a task's status.message (used for
// failed/canceled error output).
func statusMessageText(task A2ATask) string {
	msg := task.Status.Message
	if msg == nil {
		return ""
	}
	var texts []string
	for _, p := range msg.Parts {
		if p.Kind == "text" && p.Text != nil {
			texts = append(texts, *p.Text)
		}
	}
	return strings.Join(texts, "\n")
}

// originOf returns scheme://host for a URL (the card origin fallback endpoint).
func originOf(raw string) string {
	u, err := url.Parse(raw)
	if err != nil || u.Scheme == "" || u.Host == "" {
		return raw
	}
	return u.Scheme + "://" + u.Host
}

// ---------------------------------------------------------------------------
// Skill → Tool.
// ---------------------------------------------------------------------------

// taskSchema is the fixed input schema every A2A tool advertises.
func taskSchema() JSONSchema {
	return JSONSchema{
		"type": "object",
		"properties": map[string]any{
			"task": map[string]any{
				"type":        "string",
				"description": "The task to send to the agent, in natural language.",
			},
		},
		"required":             []string{"task"},
		"additionalProperties": false,
	}
}

// skillTool builds one source:"a2a" Tool for a single advertised skill.
func skillTool(agentName, endpoint string, skill A2ASkill, ag Agent) Tool {
	skillID := skill.ID
	if skillID == "" {
		skillID = skill.Name
	}
	name := Sanitize(agentName) + "_" + Sanitize(skillID)
	description := skill.Description
	if description == "" {
		description = skill.Name
	}
	if description == "" {
		description = skillID
	}
	defTimeout := ag.timeout()
	pollEvery := ag.pollEvery()

	return Tool{
		Name:        name,
		Description: description,
		InputSchema: taskSchema(),
		Source:      SourceA2A,
		Execute: func(args map[string]any, tctx *ToolContext) (result ToolResult, _ error) {
			start := time.Now()
			budget := defTimeout
			parent := context.Background()
			if tctx != nil {
				if tctx.Timeout > 0 {
					budget = time.Duration(tctx.Timeout) * time.Millisecond
				}
				if tctx.Ctx != nil {
					parent = tctx.Ctx
				}
			}
			budgetMs := int(budget / time.Millisecond)
			headers := expandHeaders(ag.Headers)
			taskText := toStr(args["task"])

			polls := 0
			taskID := ""
			state := "submitted"
			meta := func() map[string]any {
				return map[string]any{
					"agent":  agentName,
					"taskId": taskID,
					"state":  state,
					"polls":  polls,
					"ms":     int(time.Since(start).Milliseconds()),
				}
			}
			// recover() guard: no panic ever crosses the tool boundary.
			defer func() {
				if r := recover(); r != nil {
					result = ToolResult{Output: fmt.Sprintf("%v", r), IsError: true, Metadata: meta()}
				}
			}()

			// 1. SendMessage (non-blocking) → a submitted Task.
			sendParams := map[string]any{
				"message": map[string]any{
					"role":      "user",
					"messageId": uuid.NewString(),
					"parts":     []any{map[string]any{"kind": "text", "text": taskText}},
				},
				"configuration": map[string]any{"blocking": false},
			}
			raw, err := jsonRPC(parent, endpoint, "SendMessage", sendParams, headers, budget)
			if err != nil {
				return ToolResult{Output: err.Error(), IsError: true, Metadata: meta()}, nil
			}
			var task A2ATask
			_ = json.Unmarshal(raw, &task)
			taskID = task.ID
			if task.Status.State != "" {
				state = task.Status.State
			}

			// 2. Poll GetTask until terminal / timeout / cancel.
			for !terminal(state) {
				if aborted(parent) {
					state = "canceled"
					return ToolResult{Output: fmt.Sprintf("A2A task %s canceled", taskID), IsError: true, Metadata: meta()}, nil
				}
				if time.Since(start) >= budget {
					return ToolResult{
						Output:   fmt.Sprintf("A2A task %s timed out after %dms (state=%s)", taskID, budgetMs, state),
						IsError:  true,
						Metadata: meta(),
					}, nil
				}
				sleepCtx(parent, pollEvery)
				// Cancelled during the wait → stop before another GetTask.
				if aborted(parent) {
					state = "canceled"
					return ToolResult{Output: fmt.Sprintf("A2A task %s canceled", taskID), IsError: true, Metadata: meta()}, nil
				}
				raw, err := jsonRPC(parent, endpoint, "GetTask", map[string]any{"id": taskID}, headers, budget)
				if err != nil {
					// A cancellation that lands mid-request is still a cancel, not
					// a transport error (§7A: ctx abort ⇒ "A2A task <id> canceled").
					if aborted(parent) {
						state = "canceled"
						return ToolResult{Output: fmt.Sprintf("A2A task %s canceled", taskID), IsError: true, Metadata: meta()}, nil
					}
					return ToolResult{Output: err.Error(), IsError: true, Metadata: meta()}, nil
				}
				var next A2ATask
				_ = json.Unmarshal(raw, &next)
				task = next
				polls++
				if next.Status.State != "" {
					state = next.Status.State
				}
			}

			// 3. Map the terminal Task → ToolResult.
			if state == "completed" {
				return ToolResult{Output: extractOutput(task), IsError: false, Metadata: meta()}, nil
			}
			detail := statusMessageText(task)
			out := fmt.Sprintf("A2A task %s %s", taskID, state)
			if detail != "" {
				out += ": " + detail
			}
			return ToolResult{Output: out, IsError: true, Metadata: meta()}, nil
		},
	}
}

// AgentTools resolves an Agent to its tools: fetch the card, read skills[], and
// produce one Tool per skill. The agent name prefix comes from the card's name.
func AgentTools(ctx context.Context, ag Agent) ([]Tool, error) {
	if ctx == nil {
		ctx = context.Background()
	}
	timeout := ag.timeout()
	headers := expandHeaders(ag.Headers)
	card, err := fetchCard(ctx, ag.Card, headers, timeout)
	if err != nil {
		return nil, err
	}
	agentName := card.Name
	if agentName == "" {
		agentName = "agent"
	}
	// The card's url is the JSON-RPC endpoint; fall back to the card origin.
	endpoint := card.URL
	if endpoint == "" {
		endpoint = originOf(ag.Card)
	}
	tools := make([]Tool, 0, len(card.Skills))
	for _, skill := range card.Skills {
		tools = append(tools, skillTool(agentName, endpoint, skill, ag))
	}
	return tools, nil
}
